### What is Pandas ?
    -- > Pandas is a Python Library that makes data science easy. It is a tool mainly used for data manipulation and
        analysis.
    -- > In pandas we often hear a word called DataFrame. 
### What is DataFrame ?
    --> DataFrame is a 2-dimensional labeled data structure with columns of potentially different types. You can think of it like a spreadsheet or SQL table, or a dict of Series objects. It is generally the most commonly used pandas object.
        
### Creating DataFrame
Lets Understand the concept by taking an example.

            1. I have some Website Traffic Data like :
                      Day Date         Visitors    Number of Visitors liked the web page   
                      1   11/02/2019     55              40
                      2   12/02/2019     48              30
                      3   13/02/2019     69              49
                      4   14/02/2019     79              4
                      5   15/02/2019     108             73
                      6   16/02/2019     95              10
                      
            Q1. Find the Max number of visitors to the website.
            Q2. On which dates Max and Min number of visitors visited the site.
            Q3. Find the average number of visitor liked the website.
            Q4. Find the correlation between the Visitor and No. of Visitors liked the Web Page
            
            
The Above represented data is properly arranged into rows and colomns.Right ?
Now the Question comes,How can we represent the same data that every data item is very efficiently accessable and how can we able to perform some manipulation or analsis operations on the Data.

The DataFrame is the object of pandas that help with that issue. Let's create a dataFrame containing the above data.
You all might know about Python dictionaries. Lets create a Python dictionary containing the Above data........

In [1]:
WebSite_Data = {
    'Day':[1,2,3,4,5,6],
    'Date':['11/02/2019','12/02/2019','13/02/2019','14/02/2019','15/02/2019','16/02/2019'],
    'Visitors':[55,48,69,79,108,95],
    'No. Of Vistors Liked':[40,30,49,4,73,10]
}

#### Now we have a simple Python dictionary contains the required data... Now lets create a dataframe with that python dict. 
        To create a dataframe we need to import our pandas library. Right?
        import pandas as pd -> Here, pd is an alias for pandas. All the pandas classes/objects can simply be accessed by pd  instead as pandas. pd is most commonly used alias for pandas by most of the data engineers. For this session we follow  the same representation as well...

In [2]:
import pandas as pd    # Importing our library

In [3]:
df = pd.DataFrame(WebSite_Data)  

    1. DataFrame() is the method written/avaliable in pandas library.
    2. This DataFrame function takes the data as the parameter. Here the data can be of Python Dictionary, Python List or 
        Tuples as well.Here, we have passed Python Dictionary as the data....
    3. The DataFrame() Method/function creates dataframe object with the data passed. To store the returned dataframe object 
        we use 'df' variable... You can use any name, as df is also most commonly used to represent the dataframe object I 
        have used here.
    4. To view how the data is created in Dataframe object(Here df), simply we can print. But if we print the data the 
        output is not going to be good looking always. So Most of the people use head() function to see the output...

In [4]:
print("Printing the Dataframe by normal print statement of python :")
print("------------------------------------------------------------")
print(df)

Printing the Dataframe by normal print statement of python :
------------------------------------------------------------
   Day        Date  Visitors  No. Of Vistors Liked
0    1  11/02/2019        55                    40
1    2  12/02/2019        48                    30
2    3  13/02/2019        69                    49
3    4  14/02/2019        79                     4
4    5  15/02/2019       108                    73
5    6  16/02/2019        95                    10


In [5]:
print("Printing the Dataframe by using head function of pandas :")
print("df.head() function by default print the first 5 datapoints....Here to see all the data points I use df.head(10)")
print('''
        As discussed earlier head() is function avaliable in pandas. The head() function can be called by the dataframe object 
      so the statement is df.head()(dataframe_object.method())
      ''')
df.head(10)

Printing the Dataframe by using head function of pandas :
df.head() function by default print the first 5 datapoints....Here to see all the data points I use df.head(10)

        As discussed earlier head() is function avaliable in pandas. The head() function can be called by the dataframe object 
      so the statement is df.head()(dataframe_object.method())
      


Day        Date  Visitors  No. Of Vistors Liked
0    1  11/02/2019        55                    40
1    2  12/02/2019        48                    30
2    3  13/02/2019        69                    49
3    4  14/02/2019        79                     4
4    5  15/02/2019       108                    73
5    6  16/02/2019        95                    10

#### So far we have completed creating the dataframe for storing the required data....  Now before Answering the above question lets see how to access the elements....

In [6]:
print("Shape of our DataFrame :",df.shape)
print("The datatypes of the elements :")
print(df.dtypes)
print("\nThis means the Day column contains of int64 data elements, Date contains object(Strings) type data elements and so on....")

Shape of our DataFrame : (6, 4)
The datatypes of the elements :
Day                      int64
Date                    object
Visitors                 int64
No. Of Vistors Liked     int64
dtype: object

This means the Day column contains of int64 data elements, Date contains object(Strings) type data elements and so on....


#### -->How to Print only one Column. (Say Print Visitors Colomn....)
            1. There are two ways of printing column data...
                i) df.column_name
                ii) df['column_name']
             -> Which type is most preferable and why?

In [7]:
df.Visitors # Prints Visitors column data

0     55
1     48
2     69
3     79
4    108
5     95
Name: Visitors, dtype: int64

In [8]:
df['Visitors']

0     55
1     48
2     69
3     79
4    108
5     95
Name: Visitors, dtype: int64

##### From the above two which method is most preferable and why??? 
        1. The second menthod of retrieving data is most preferable one.
            i) Lets take an instance, consider the column 4(No. Of Vistors Liked). You cannot retrive the data of that 
                column by using the first method. Since we have spaces in the name of the column.
            ii) To get the data of two or more columns at a time first method is no logner valid..... you cannot get two 
                columns data at one instance of time...
                
     Lets print Two columns data... Visitors and No. Of Vistors Liked....

In [9]:
df[['Visitors','No. Of Vistors Liked']]

Visitors  No. Of Vistors Liked
0        55                    40
1        48                    30
2        69                    49
3        79                     4
4       108                    73
5        95                    10

        --->So you might have understand how to get data form a single column or multiple columns... Now lets look how to get data through rows....

In [10]:
# To get data from first row...
df.head(1)

Day        Date  Visitors  No. Of Vistors Liked
0    1  11/02/2019        55                    40

In [11]:
# To get data of first 3 rows...
df.head(3)

Day        Date  Visitors  No. Of Vistors Liked
0    1  11/02/2019        55                    40
1    2  12/02/2019        48                    30
2    3  13/02/2019        69                    49

In [12]:
# To get data from last 2 rows....
df.tail(2)

Day        Date  Visitors  No. Of Vistors Liked
4    5  15/02/2019       108                    73
5    6  16/02/2019        95                    10

In [13]:
# To get data from all rows.....
df[:]

Day        Date  Visitors  No. Of Vistors Liked
0    1  11/02/2019        55                    40
1    2  12/02/2019        48                    30
2    3  13/02/2019        69                    49
3    4  14/02/2019        79                     4
4    5  15/02/2019       108                    73
5    6  16/02/2019        95                    10

In [14]:
# Get data From 3rd index..... Here Index starts form 0
df.loc[3]

Day                              4
Date                    14/02/2019
Visitors                        79
No. Of Vistors Liked             4
Name: 3, dtype: object

In [15]:
# Get names of our columns.......
df.columns

Index(['Day', 'Date', 'Visitors', 'No. Of Vistors Liked'], dtype='object')

#### We can discuss more about the basics and all the different stuff in later sessions.... Lets end this session by answering Our Questions.......
        Q1. Find the Max number of visitors to the website.
        Q2. On which dates Max and Min number of visitors visited the site.
        Q3. Find the average number of visitor liked the website.
        Q4. Find the correlation between the Visitor and No. of Visitors liked the Web Page

### Answer 1:

In [16]:
print("Answer#1 : Max number of visitors to the site :",df['Visitors'].max())

Answer#1 : Max number of visitors to the site : 108


        --> To get the max value in the required column, we can simply get the max value by specifing the column that we 
            need and applying the max() function on it...
        --> In the similar fasion, We can get the Min, Mean, Std, median, cumulative sum values as well....

In [17]:
print("Min number of visitors to the site :",df['Visitors'].min())
print("Mean of Visitors :",df['Visitors'].mean())
print("Std of Visitors :",df.Visitors.std())

Min number of visitors to the site : 48
Mean of Visitors : 75.66666666666667
Std of Visitors : 23.096897338531566


### Answer 2:

In [20]:
print("Answer#2 :")
print("\t\tDate of Max no. of Visits :\t\t",df['Date'][df['Visitors']==df['Visitors'].max()] )
print("\t\tDate of Min no. of Visits :\t\t",df['Date'][df['Visitors']==df['Visitors'].min()])

Answer#2 :
		Date of Max no. of Visits :		 4    15/02/2019
Name: Date, dtype: object
		Date of Min no. of Visits :		 1    12/02/2019
Name: Date, dtype: object


        ---> The output for max no. of visits is : 15/02/19 that is True
        ---> The output for min no. of visits is : 12/02/19 that is True

#### df['What we Need'][Condition]....... This is what we have used....
        
        df['Date'][df['Visitors']==df['Visitors'].max()]
            ----> Here We need the date as our output... df['Date'] will print/return all the dates....
                  now we need only one date when the traffic(Visitors) are high/max..
                  df['Visitors'] == df['Visitors'].max() - is our condition. This Will stand True when the Visitors Are High
            ----> In words our statement mean :
                    Print the date in which the condition stand True....
                  

#### Question 2 Extended: 
                Print the no of visits also with the date.

In [4]:
df[['Date','Visitors']][df['Visitors']==df['Visitors'].max()]

Date  Visitors
4  15/02/2019       108

In [5]:
df[['Date','Visitors']][df['Visitors']==df['Visitors'].min()]

Date  Visitors
1  12/02/2019        48

### Comparing with------- >df['What we Need'][Condition]
        ---> Here What we need is Date and Visitors.....
                So, df[['Date','Visitors']] - This will give both Date and Visitors columns completely.... To get only 
                required quatity we are placing the condition next to it....

### Answer 3:

In [31]:
print("Average No. of likes to the Page/Site :(Simply Finding Mean)",df['No. Of Vistors Liked'].mean())
print("Average No. of likes to the Page/Site :(Mathematical Calculation)",sum(df['No. Of Vistors Liked'])/len(df))

Average No. of likes to the Page/Site :(Simply Finding Mean) 34.333333333333336
Average No. of likes to the Page/Site :(Mathematical Calculation) 34.333333333333336


### Answer 4:

In [36]:
df.drop(['Day'],1).corr()

Visitors  No. Of Vistors Liked
Visitors              1.000000              0.197529
No. Of Vistors Liked  0.197529              1.000000

        --> Correlation : It gives a mutual relationship or connection between two or more things.
            Here : Visitors comparing with itself Completely overlap... All the entities completely match with each other
                   Hence, 1st Value here is 1.000000.
                   Now, Visitors comparing with No. Of Visitors Liked The value is 0.197529 indicating both are very loosly
                   Related....
#### This type of analizing relationship, we are going to use more in Machine Learning....

### Hope this notebook was Helpful !!! Thank You...

##### Notebook Prepared By : V.PoornaChand
https://www.linkedin.com/in/poorna-chand-902642177/